### ISSUE TITLE
Kubelet evictions - whats remaining?

### ISSUE THREAD

**Issue by vishh** [kubernetes/kubernetes#31362]
Kubelet now supports evicting of pods & images to free up memory and disk space whenever it notices memory or disk pressure. This feature is amazing since it helps keep the nodes running. But there are still a few glitches with this feature that makes it not ready for production. This issue attempts to expose those glitches and discuss possible solutions for those glitches.
- Kubelet pod evictions are not checkpointed. Kubelet restarts might forget evicted pods. need to checkpoint evicted phase to the API server before performing actual eviction
- Kubelet eviction triggers are slow. It takes at-least 10 seconds for kubelet to detect resource usage changes. In the case of disk usage, it can be as long as 70 seconds. This makes low thresholds for kubelet evictions not work in production. For example `memory.available<100Mi`, doesn't prevent system OOMs ~40% of the time. This issue is more serious of disk.
  Proposed Solutions
  - Add an on-demand stats interface to cAdvisor that collects & provides the most recent stats on-demand - https://github.com/google/cadvisor/issues/1247 is related
  - Setup memcg thresholds that trigger events as soon as the eviction thresholds are met.
- Kubelet evicts more pods that needed whenever it notices memory or disk pressure. 
  - Memory - the 10 second update could result in multiple evictions. Kubelet evicts volumes asynchronously after evicting pods. This results in memory held my in-memory volumes from not being released and additional pods get deleted. Possible solutions here include evictions volumes before marking a pod as "evicted" and running eviction control loops more often (1s?) whenever there is pressure.
  - Disk - The issue mentioned above from memory applies to disk too. In addition to that, the disk storage layout has not been spec'ed yet - https://github.com/kubernetes/kubernetes/issues/30799
    Disk Inodes are not tracked per-pod. This could result in eviction multiple pods until the outlier is evicted.
    Logs are not preserved for evicted pods. This might be critical for certain applications.

cc @kubernetes/sig-node @derekwaynecarr @sjenning @ronnielai

**Comment by derekwaynecarr** [comment]
>few glitches with this feature that do not make it ready for production

I disagree with the above statement.  I would strongly recommend user's to enable this feature in production, but I also agree setting an eviction threshold at `memory.available<100Mi` is skating on thin-ice with the current support.  I know per our conversation, @vishh feels similarly, but I want to clarify the statement for others that come and read this issue later.  Many operators of production workloads target 70% utilization, and run pods with requests and limits both specified within a well-defined level of over-commit.  Those users can derive significant value from the existing support and we should encourage our users to enable the function to reduce their risk of OOM, if not completely eliminate it yet.

That said, I agree we should focus on the following areas:

**Improve the latency in observing when eviction thresholds are met**

For `disk`, we will need to continue to poll `cAdvisor`, and since image garbage collection is unified with `disk` management, we should not promote user's running so close to total out of disk scenarios.

For `memory`, we should integrate with `memcg` notification API in addition to polling `cAdvisor`.  For each soft/hard threshold for memory, we should register notifications to get a call-back.  This will give a much more precise mechanism to enforce grace-period for soft evictions, and vastly reduce the latency in observing memory pressure.  Since observation of memory pressure is now disconnected from knowing how much memory each pod/container is consuming in this scenario, we need to improve `cAdvisor` stat collection so we make the right eviction decision in response.  There is reduced value to know there is a problem, and then solve it by selecting an incorrect pod.

**On-demand stats when ranking pods for eviction**

`cAdvisor` should support on-demand stat collection, and the eviction manager should request on-demand usage stats prior to making a pod eviction decision.  If we are unable to get on-demand stats complete in 1.5 time-frame, we should at minimum expose a global timestamp to prevent excessive pod deletion due to eviction manager looking at the same usage stat values in consecutive sync passes.

**Pod evictions should block on volume clean-up**

If `disk` is scarce, pod evictions should block on per pod volume clean-up rather than happen asynchronously.  For `memory`, we may not want to block since under pressure.

**Inode reclamation**

Right now, we do not have per-container usage stats for inodes, which means knowing which pod to reclaim in the face of inode exhaustion is difficult.  In practice though, inode exhaustion would also trigger image and container gc.  Right now, we have no visibility into the number of inodes we reclaim in response to those actions.  Rather than prioritizing per container stats for inode usage, we would prefer to prioritize on-demand stat collection in `cAdvisor` so we do not err on the side of an unnecessary pod eviction after image/container gc.

**Comment by sjenning** [comment]
Doing some research on how to get better (i.e. low latency) notification of memory pressure.  There is a memcg attribute ```memory.pressure_level``` which will give the following notifications when the cgroup is under pressure:
- low: system is reclaiming "inactive" memory for new allocations
- medium: system is reclaiming "active" memory which might include swap activity
- critical: the amount of reclaimable memory is very low and the cgroup is thrashing.  OOM is imminent.

The medium-level notification on the root cgroup might be a great indicator of when pods should start being evicted.  It seems that is the point at which the overall performance is beginning to suffer due to reclamation of active cache memory.  I'll have to see how easily this is triggered though.

**Comment by vishh** [comment]
I suspect that by the time kubelet reacts to `medium` pressure, there could
be one or more system OOM kills. If we were to run a dedicated OOM daemon,
then I'd be less concerned about using these notifications.
For now, I prefer using `memory thresholds` instead. WDYT?

**Comment by sjenning** [comment]
Yes, I've been doing some tests and, especially for systems with no swap (or ```memory.memsw.limit_in_bytes``` equal to ```memory.limit_in_bytes```), there doesn't seem to be a notification level that is useful :frowning: ```low``` is too easily triggered (happens on any cache reclamation) and ```medium``` and ```critical``` happen too late to be of any use.  If you have swap, then medium is somewhat useful, but that can't be assumed.

```memory.pressure_level``` is a dead end.

**Comment by vishh** [comment]
We recommend disabling swap since it results in unpredictable application
performance in a cluster scenario.

**Comment by ravilr** [comment]
@derekwaynecarr @vishh  in regards to imagefs garbage collection, if i want to set a lower threshold of  say '<50%', would the node status condition be updated as 'under disk pressure', thereby rejecting new pods ? I think disk cleanup configured at lower threshold shouldn't lead to scheduler disregarding the node as schedulable or kubelet admission controlling the pods.
what i'm saying is people would want the kubelet to kick in garbage collection way before than any pod eviction thresholds. Is that possible now?

**Comment by ravilr** [comment]
my thinking is around the scenario that i would like to avoid, where most of the nodes in the cluster reached the disk based eviction threshold at the same time and become unschedulable till the usage falls below the threshold. I understand that existing pods wouldn't be evicted before running the cleanup, but new pods wouldn't be admitted for the eviction pressure transition period.  
Also, certain types of disk performance may degrade at fuller capacity, so one would want to avoid filling up near-to-full, if possible.
Also in continous deployment world, where every application change results in a new image getting deployed,  it could be that lot of the images are of older versions and not enough incentive to keep them lying around on the disk.

**Comment by vishh** [comment]
If images are the main issue, by setting a lower hard threshold (~70%) and a lower eviction transition period (~30s), the nodes can be made to evict images often and not end up staying unschedulable for too long. This will result in lowering effective utilization of disk to 70% though.

One possibility is to extend soft evictions to start performing image and logs garbage collection as soon as the threshold is met and evict pods only after the soft eviction graceperiod is met. We can also provide an option to disable pod evictions via soft threshold by setting the soft eviction grace period to `-1` or some such special value. @derekwaynecarr WDYT?

**Comment by derekwaynecarr** [comment]
@ravilr - I am curious what storage driver you are using?

**Comment by ravilr** [comment]
@derekwaynecarr - using devicemapper driver.

**Comment by derekwaynecarr** [comment]
@ravilr - do you use the autoextend features?  If so with what percent?
Also how do you configure dm.min_free_space?

@vishh - I am not adverse doing node level reclaims the moment an soft
threshold is passed, but I need to reason through some edge cases.

**Comment by ravilr** [comment]
@derekwaynecarr - autoextend is turned off, with thinpool DATA_SIZE set to 90%FREE. dm.min_free_space is also not set. i guess the default is 10%. Do you suggest configuring differently and why? Thanks.

**Comment by dims** [comment]
This needs to be triaged as a release-blocker or not for 1.5 @vishh @derekwaynecarr

**Comment by davidopp** [comment]
Eviction for overloaded CPU would probably be good

**Comment by vishh** [comment]
@davidopp Why do you think so? CPU is compressible right?

**Comment by davidopp** [comment]
To spread out the load if best effort pods are getting starved. I guess rescheduler could do it instead...

**Comment by derekwaynecarr** [comment]
@davidopp - we had discussed doing eviction on cpu load, but i also think the rescheduler could take on that responsibility as well... we can discuss more in resource management group.

**Comment by dashpole** [comment]
I have a few issues regarding evicting extra pods:
#41347
#40474

**Comment by dashpole** [comment]
@vishh, should we move this to 1.7?  I think all of the eviction work for 1.6 are completed.

**Comment by jayunit100** [comment]
There is at least some user feedback requesting CPU eviction that parallels what we're doing for memory.  is that something we agree on ?

**Comment by vishh** [comment]
@dashpole can you work on a list of work items that are pending based on this issue?

**Comment by dashpole** [comment]
Current state of kubelet evictions:
 - Memory Evictions: With fairly well-behaved memory-consuming pods ([using 12Mi memory every 10 sec](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/memory_eviction_test.go#L280)), there is no evidence of extra pod evictions.  Once other, more pressing issues have been dealt with, it may be worth testing with more aggressive memory consuming pods.  Enabling memcg thresholds should allow the kubelet to handle these more aggressive pods without extra evictions.
 - Disk Evictions: Stats collection is slow, and stats used for evictions are frequently stale.  This is the main blocker to reducing extra evictions.

My working items for 1.7 with links to issues: 
 - On Demand Stats [google/cadvisor#1247](https://github.com/google/cadvisor/issues/1247)
 - Test, and eventually reenable memcg on GCI #42676 
 - Optimize container disk stats collection: [google/cadvisor#1576](https://github.com/google/cadvisor/pull/1576)
 - Collect and expose metrics on age of stats used for evictions #42923
 - Collect and expos metrics for latency for high-latency stats collection [google/cadvisor#1610](https://github.com/google/cadvisor/issues/1610)
 - Fix Pod Eviction Race condition #40474

Other items mentioned here that require more discussion, and an owner:
 - CPU based evictions #42910 
 - Eviction Checkpointing
 - Blocking evictions on volume cleanup

**Comment by dashpole** [comment]
Another working item for 1.7:
#43166

**Comment by dashpole** [comment]
cc/ @dchen1107

**Comment by dashpole** [comment]
Update for 1.7 items:
On Demand Stats: Punted to next release
Reenable memcg on GCI: **On track to finish** by 1.7
Optimize container disk stats collection: More complicated than we initially thought.
Metrics on age of status used for evictions: **Complete**, metrics are exposed, and logged during eviction testing.
Metrics for high-latenct stats collection: **Planned to complete for 1.7**
Fix or Mitigate Pod Eviction Race Condition:  **Complete**
Eviction waits for pod cleanup: #43166 **Complete**
Garbage Collection of dead containers under disk pressure: #45896 **Pr out for review**
Set default thresholds for disk eviction: **Planned to complete for 1.7**

**Comment by jayunit100** [comment]
Can I squeeze https://github.com/kubernetes/kubernetes/pull/45141 for 1.7?  This will allow users to reason about the NodeNotReady error in a much more transparent manner.  It's non functional as a modification.

**Comment by dchen1107** [comment]
We made a lot improvements on kubelet eviction for 1.7 release, and reenable memcg notification on GCI is tracked by a separate issue. 

I move this umbrella issue to 1.8 for more improvement.

**Comment by dashpole** [comment]
There are a couple items I still would like to tackle related to evictions in the 1.10 - 1.11 timeframe:
Enable On-Demand metrics: https://github.com/kubernetes/kubernetes/pull/56112
Fix Implementation of Memory Cgroup notifications: https://github.com/kubernetes/kubernetes/issues/51745, https://github.com/kubernetes/community/pull/1451
Cleanup eviction signals: https://github.com/kubernetes/kubernetes/issues/53902
Prevent evictions after successfully performing garbage collection: https://github.com/kubernetes/kubernetes/issues/56573
Monitor allocatable memory usage more directly: https://github.com/kubernetes/kubernetes/issues/55638
Use Memory Cgroup notifications for allocatable evictions: https://github.com/kubernetes/kubernetes/issues/57901

/remove-lifecycle stale

**Comment by dashpole** [comment]
The remaining items, to be finished in 1.11 are:
Use Memory Cgroup notifications for allocatable evictions: https://github.com/kubernetes/kubernetes/issues/57901
Better kubelet eviction events: https://github.com/kubernetes/kubernetes/pull/63415
Eviction tests ensure `Status.Reason` is `Evicted` so we can catch OOMs: https://github.com/kubernetes/kubernetes/pull/57849

**Comment by dashpole** [comment]
Everything I had planned to work on for this has been completed.
To address the initial points, we have long had per-container inode metrics and inode eviction, we have on-demand node-level and allocatable-level metrics, a decent integration with memory cgroup notifications, and do not perform additional evictions until all artifacts of the previous evicted pod, including volumes and containers, have been cleaned up.
/close
